### PR TITLE
If

### PR DESCRIPTION
**Pull request by MarcelGarus**

### CODE REVIEW

File: packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart [hunk 1]
```diff
@@ -142,6 +163,14 @@ class IdFinderVisitor extends hir.ExpressionVisitor<Option<hir.Expression>> {
     if (node.id == id) return Some(node);
     return None();
   }
+
+  @override
+  Option<hir.Expression> visitAssignmentExpression(AssignmentExpression node) {
+    if (node.id == id) return Some(node);
+    node.right.accept(this);
+
+    return None();
```
Comment (JonasWanke) [inline packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart]:
```suggestion
    return node.right.accept(this);
```

File: packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart [hunk 2]
```diff
@@ -740,6 +776,32 @@ class ExpressionContext extends InnerContext {
...
+  IfContext(
+    Context parent,
+    this.id,
+    this.label,
+  )   : assert(id != null),
+        assert(label != null),
+        super(parent);
+
+  final DeclarationLocalId id;
+  final Option<String> label;
```
Comment (JonasWanke) [inline packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart:789]:
```suggestion
  IfContext(
    Context parent,
    this.id,
  )   : assert(id != null),
        super(parent);

  final DeclarationLocalId id;
```

File: packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart [hunk 3]
```diff
@@ -1030,6 +1092,48 @@ extension on Context {
...
+    final elseBody = <Expression>[];
+    if (theIf.elseKeyword != null) {
+      assert(theIf.elseBody != null);
+      final thenContext = IfContext(this, getId(theIf), None());
+      final loweredThenBody = theIf.thenBody.expressions.map((expression) {
+        return thenContext
+            .innerExpressionContext(forwardsIdentifiers: true)
+            .lowerUnambiguous(expression);
+      }).merge();
+      if (loweredThenBody is Error) return loweredThenBody;
```
Comment (JonasWanke) [inline packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart]:
```suggestion
      final elseContext = IfContext(this, getId(theIf), None());
      final loweredElseBody = theIf.elseBody.expressions.map((expression) {
        return elseContext
            .innerExpressionContext(forwardsIdentifiers: true)
            .lowerUnambiguous(expression);
      }).merge();
      if (loweredElseBody is Error) return loweredElseBody;
```

File: packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart [hunk 4]
```diff
@@ -1030,6 +1092,48 @@ extension on Context {
...
+      elseBody.addAll(loweredThenBody.value);
+    }
+
+    final type = expressionType.valueOrNull ??
+        hir.CandyType.union([
+          if (thenBody.isNotEmpty) thenBody.last.type,
+          if (elseBody.isNotEmpty) elseBody.last.type,
+        ]);
```
Comment (JonasWanke) [inline packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart]:
```suggestion
        hir.CandyType.union({
          if (thenBody.isNotEmpty) thenBody.last.type,
          if (elseBody.isNotEmpty) elseBody.last.type,
        }.toList());
```

File: packages/compiler_dart/lib/src/body.dart [hunk 5]
```diff
@@ -206,6 +214,25 @@ class DartExpressionVisitor extends ExpressionVisitor<List<dart.Code>> {
           dart.Code('return;'),
       ];
 
+  @override
+  List<dart.Code> visitIfExpression(IfExpression node) {
+    List<dart.Code> visitBody(List<Expression> body) => [
+          for (final expression in body) ...expression.accept(this),
+          if (body.isNotEmpty && body.last.type != CandyType.unit)
```
Comment (JonasWanke) [inline packages/compiler_dart/lib/src/body.dart]:
```suggestion
          if (body.isNotEmpty && node.type != CandyType.unit)
```

File: packages/compiler_dart/lib/src/builtins.dart [hunk 6]
```diff
@@ -77,54 +77,54 @@ abstract class BuiltinCompiler<Output> {
 class DartBuiltinCompiler extends BuiltinCompiler<dart.Spec> {
   @override
   Option<dart.Spec> compileAny() {
-    // `Any` corresponds to `Object`, hence nothing to do.
+    // `Any` corresponds to `Object`, hence Never to do.
```
Comment (JonasWanke) [inline packages/compiler_dart/lib/src/builtins.dart]: Uhm…

File: packages/parser/lib/src/parser/grammar.dart [hunk 7]
```diff
@@ -659,9 +659,9 @@ class ParserGrammar {
             _id++,
             ifKeyword: left[0] as IfKeywordToken,
             condition: value,
-            thenExpression: right[1] as Expression,
+            thenBody: right[1] as LambdaLiteral,
             elseKeyword: elsePart?.elementAt(1) as ElseKeywordToken,
-            elseExpression: elsePart?.elementAt(3) as Expression,
+            elseBody: elsePart?.elementAt(3) as LambdaLiteral,
```
Comment (JonasWanke) [inline packages/parser/lib/src/parser/grammar.dart:664]: Do we want to constrain those to lambda-blocks? Earlier, we also allowed something like `if (true) a else b`

**Review by JonasWanke — CHANGES_REQUESTED** [review]

File: packages/parser/lib/src/parser/grammar.dart (hunk 7 above)
Comment (MarcelGarus) [reply packages/parser/lib/src/parser/grammar.dart:664]: We'll leave this as it is for now and think about when if-syntax we allow when we write the new parser.

File: packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart (hunk 2 above)
Comment (MarcelGarus) [reply packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart:789]: We decided to leave the label and instead catch returns targeted at ifs.

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**